dod/src/main/java/com/revolut/rxdata/dod/SharedSingleRequest.kt

```diff
@@ -19,15 +19,22 @@ import io.reactivex.Single
  *
  */
 
-internal class SharedSingleRequest<Params, Result : Any>(
+internal class SharedSingleRequest<Params: Any, Result : Any>(
     private val load: (params: Params) -> Single<Result>
 ) {
 
     private val sharedObservableRequest: SharedObservableRequest<Params, Result> =
         SharedObservableRequest { params ->
-            load(params).toObservable()
+            load(params)
+                .doOnSuccess { removeRequest(params) }
+                .doOnError { removeRequest(params) }
```

Comment:
I guess we need to call `removeRequest` at dispose as well (and yes, that removal should be more sophisticated and actually check a correspondence of the deleted request to the current one, or enforce somehow **At Most Once** behaviour for `removeRequest`)

there is test which checks that especial issue:
```
    @Test
    fun `GIVEN network not answer before unsubscription of first observe WHEN observe again THEN second observer see network response`() {
        // given
        whenever(fromNetwork.invoke(any())).thenReturn(Single.never())
        storage.remove(params)
        memCache.remove(params)

        val observer1 = dataObservableDelegate.observe(params = params, forceReload = false)
            .skipWhileLoading()
            .test()
        ioScheduler.triggerActions()

        observer1.dispose()
        observer1.assertValues()

        // when
        whenever(fromNetwork.invoke(any())).thenReturn(Single.just(domain))
        val observer2 = dataObservableDelegate.observe(params = params, forceReload = false)
            .skipWhileLoading()
            .test()
        ioScheduler.triggerActions()

        // then
        observer2.assertValues(Data(domain))    // ACTUAL: EMPTY LIST
    }
    ```

Reply:
We do remove request on dispose, it's handled in SharedObservableRequest class via `doFinally { removeRequest(params) }`. here we only add extra removals on events that are specific to Single lifecycle - i.e. doOnSuccess, which doesn't exist in Observable lifecycle. Actually even doOnError is redundant. 

Your test case fails because DoD has internal timeout logic (60s by default ) and it's keeping Single.never() alive during this period. Observer2 basically resubscribes to this request in this case. If you change it slightly to accomodate for timeouts: 

```
        // when
        whenever(fromNetwork.invoke(any())).thenReturn(Single.just(domain))
        ioScheduler.advanceTimeBy(180, TimeUnit.SECONDS)

        val observer2 = dataObservableDelegate.observe(params = params, forceReload = false)
            .skipWhileLoading()
            .test()
        ioScheduler.triggerActions()
```
then it passes. 

It's covered in `WHEN unsubscribed from network and data NOT arrives in 60 seconds THEN data not saved` so I won't be adding this case

Reply:
got it, thanks